src/Authenticator/SessionAuthenticator.php

```diff
@@ -115,4 +117,68 @@ public function clearIdentity(ServerRequestInterface $request, ResponseInterface
             'response' => $response,
         ];
     }
+
+    /**
+     * Impersonates a user
+     *
+     * @param ServerRequestInterface $request
+     * @param ResponseInterface $response
+     * @param \ArrayAccess|array $impersonator
+     * @param \ArrayAccess|array $impersonated
+     * @return \ArrayAccess|array
+     */
+    public function impersonate(ServerRequestInterface $request, ResponseInterface $response, \ArrayAccess|array $impersonator, \ArrayAccess|array $impersonated): \ArrayAccess|array
+    {   $sessionKey = $this->getConfig('sessionKey');
+        $impersonateSessionKey = $this->getConfig('impersonateSessionKey');
+        /** @var \Cake\Http\Session $session */
+        $session = $request->getAttribute('session');
+        if ($session->check($impersonateSessionKey)) {
+            //throw new UnauthorizedException('You are already impersonating a user.');
```

Comment:
I think it reasonable to overwrite the existing impersonation if another is started.

Reply:
Makes sense

Reply:
Wait, doing some tests I found this use case:

* User A impersonates User B
* User B then impersonates User A
* User B (as User A) stops impersonation
* User B is logged in and no impersonation trace

So maybe we should not allow if user is impersonating already OR another way would be checking if user is being impersonated, and dont replace impersonator data but only impersonated one.

What do you think?

Reply:
I don't think multi-level / nested impersonation should be allowed. It will make thinks overly complex.

Reply:
Agree that we should only support a single level of impersonation. The scenarios where I've used impersonation in the past are in a developer/support person needing to view a user's account as that user. I've not ever needed nested impersonation in that scenario.